src/components/implementation/cbuf_mgr/default/cbuf_mgr.c

```diff
@@ -190,86 +215,148 @@ cbuf_comp_info_bin_get(struct cbuf_comp_info *cci, int sz)
 }
 
 static struct cbuf_bin *
-cbuf_comp_info_bin_add(struct cbuf_comp_info *cci, int sz)
+cbuf_comp_info_bin_add(struct cbuf_comp_info *cci, unsigned long sz)
 {
-	if (sz == CBUF_MAX_NSZ) return NULL;
 	cci->cbufs[cci->nbin].size = sz;
 	cci->nbin++;
 
 	return &cci->cbufs[cci->nbin-1];
 }
 
 static int
-cbuf_map(spdid_t spdid, vaddr_t daddr, void *page, int size, int flags)
+cbuf_map(spdid_t spdid, vaddr_t daddr, void *page, unsigned long size, int flags)
 {
-	int off;
-	assert(size == (int)round_to_page(size));
+	unsigned long off;
+	assert(size == round_to_page(size));
 	assert(daddr);
 	assert(page);
 	for (off = 0 ; off < size ; off += PAGE_SIZE) {
 		vaddr_t d = daddr + off;
-		if (d != (mman_alias_page(cos_spd_id(), ((vaddr_t)page) + off,
-						spdid, d, flags))) {
-			printd("couldn't alias(%d, %x, %d, %x, %d)\n", cos_spd_id(), ((vaddr_t)page) + off, spdid, d, flags);
-			assert(0); /* TODO: roll back the aliases, etc... */
+		if (unlikely(d != (mman_alias_page(cos_spd_id(), ((vaddr_t)page) + off,
+						   spdid, d, flags)))) {
+			for(d = daddr+off-PAGE_SIZE; d >= daddr; d -= PAGE_SIZE) {
+				mman_revoke_page(spdid, d, 0);
+			}
+			return -ENOMEM;
 		}
 	}
 	return 0;
 }
 
+/* map the memory from address p and size sz to the component spdid with 
+ * permission flags. if p is NULL allocate a piece of new memory
+ * return spdid's address to daddr, manager's virtual address to page*/
 static int
-cbuf_alloc_map(spdid_t spdid, vaddr_t *daddr, void **page, int size)
+cbuf_alloc_map(spdid_t spdid, vaddr_t *daddr, void **page, void *p, unsigned long sz, int flags)
 {
-	void *p;
 	vaddr_t dest;
 	int ret = 0;
+	void *new_p;
+
+	u64_t start, end;
+	rdtscll(start);
+
+	assert(sz == round_to_page(sz));
+	if (!p) {
+		new_p = page_alloc(sz/PAGE_SIZE);
+		assert(new_p);
+		memset(new_p, 0, sz);
+	} else {
+		new_p = p;
+	}
 
-	assert(size == (int)round_to_page(size));
-	p = page_alloc(size/PAGE_SIZE);
-	if (!p) goto done;
-	memset(p, 0, size);
-
-	dest = (vaddr_t)valloc_alloc(cos_spd_id(), spdid, size/PAGE_SIZE);
-	if (!dest) goto free;
-
-	if (!cbuf_map(spdid, dest, p, size, MAPPING_RW)) goto done;
+	dest = (vaddr_t)valloc_alloc(cos_spd_id(), spdid, sz/PAGE_SIZE);
+	if (unlikely(!dest)) goto free;
+	if (!cbuf_map(spdid, dest, new_p, sz, flags)) goto done;
 
-free:
+free: 
 	if (dest) valloc_free(cos_spd_id(), spdid, (void *)dest, 1);
-	dest = 0;
-	page_free(p, size/PAGE_SIZE);
-	p = 0;
+	if (!p) page_free(new_p, sz/PAGE_SIZE);
 	ret = -1;
-done:
-	*page  = p;
+
+done:	
+	if (page) *page  = new_p;
 	*daddr = dest;
+	rdtscll(end);
+	op_nums[7]++;
+	per_total[7] += (end-start);
```

Comment:
I'd put this op_nums and per_total in an inline function that does nothing if not DEBUG mode. e.g. op_end(CBUF_ALLOC_MAP) to increment the op_nums[CBUF_ALLOC_MAP] and per_total[...] (assuming you add enum or macros for CBUF_ALLOC_MAP).